### PR TITLE
Zombiehands

### PR DESCRIPTION
**Pull request by MTattrie**
added inventory to zombie, and method for zombie attack
added melee attack collision handler 
added weapon zombiehand
handle turret and barricades health reaching 0

TODO
apply firerate to zombies (need to talk to rob about zombie attack state)
add zombie attack detection for the base (object manager, base id is always 0?)
handle Marine death (issue)
*camera is currently tied to a marine

### CODE REVIEW

File: src/creeps/Zombie.cpp [hunk 1]
```diff
@@ -279,3 +280,19 @@ string Zombie::generatePath(const float xStart, const float yStart,
...
+/**
+ * Date: Mar. 28, 2017
+ * Author: Mark Tattrie
+ * Function Interface: void Zombie::zAttack()
+ * Description:
+ * Calls the zombies current weapon "ZombieHands" to fire
+ */
+void Zombie::zAttack(){
+    Weapon* w = inventory.getCurrent();
+    if( w != nullptr){
```
Comment (isaacmorneau) [inline src/creeps/Zombie.cpp]: just `if (w) {` is fine

File: src/inventory/Inventory.cpp [hunk 2]
```diff
@@ -5,7 +5,7 @@
 #include "../game/GameManager.h"
 #include "../log/log.h"
 
-Inventory::Inventory(): defaultGun(GameManager::instance()->generateID()) {
+Inventory::Inventory(): defaultGun(GameManager::instance()->generateID()), tempZombieHand(GameManager::instance()->generateID()) {
```
Comment (isaacmorneau) [inline src/inventory/Inventory.cpp]: wrap line

File: src/inventory/weapons/MeleeWeapon.cpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+ * Function Interface: MeleeWeapon::MeleeWeapon(string type, TEXTURES sprite, string fireSound,
+ *       string hitSound, string reloadSound, string emptySound, int range, int damage, int AOE,
+ *       int penetration, int clip, int clipMax, int ammo, int reloadDelay, int fireDelay, int32_t id)
+ * Description:
+ * Ctor for MeleeWeapon
+ */
+MeleeWeapon::MeleeWeapon(string type, TEXTURES sprite, string fireSound, string hitSound, string reloadSound,
+        string emptySound, int range, int damage, int AOE, int penetration, int clip,
+        int clipMax, int ammo, int reloadDelay, int fireDelay, int32_t id)
+: Weapon(type, sprite, fireSound, hitSound, reloadSound, emptySound, range, damage, AOE,
```
Comment (isaacmorneau) [inline src/inventory/weapons/MeleeWeapon.cpp]: spacing

File: src/inventory/weapons/MeleeWeapon.cpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    hitTurrets = ch.detectMeleeCollision(ch.getQuadTreeEntities(ch.quadtreeTurret, &mov),&mov, hitBox);
+    for(auto& y: hitTurrets){//update each zombie's helth
+        y->collidingProjectile(damage);
+        logv("hit a turret, turret > 0 = %d\n", dynamic_cast<Turret *>(y)->healthCheckTurret());
+    }
+
+    hitBarricades = ch.detectMeleeCollision(ch.getQuadTreeEntities(ch.quadtreeBarricade, &mov),&mov, hitBox);
+    for(auto& z: hitTurrets){//update each zombie's helth
```
Comment (isaacmorneau) [inline src/inventory/weapons/MeleeWeapon.cpp]: i think all preceding for s can be `const auto&` no?

File: src/inventory/weapons/MeleeWeapon.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+#include "../../collision/CollisionHandler.h"
+
+#define Z_WIDTH 75
+#define Z_HEIGHT 125
+
+class MeleeWeapon: public Weapon {
+public:
+    MeleeWeapon(string type, TEXTURES sprite, string fireSound, string hitSound, string reloadSound,
+            string emptySound, int range, int damage, int AOE, int penetration, int clip,
+            int clipMax, int ammo, int reloadDelay, int fireDelay, int32_t id);
```
Comment (isaacmorneau) [inline src/inventory/weapons/MeleeWeapon.h]: const parameters

**Review by isaacmorneau — CHANGES_REQUESTED** [review]

File: src/collision/CollisionHandler.cpp [hunk 6]
```diff
@@ -172,6 +172,26 @@ void CollisionHandler::detectLineCollision(TargetList &targetList, const int gun
...
 /**
+ * Date: Mar. 28, 2017
+ * Author: Mark Tattrie
+ * Function Interface: std::vector<Entity *> CollisionHandler::detectMeleeCollision(
+ *      std::vector<Entity*> returnObjects, const Entity *entity, const HitBox hb)
+ * Description:
+ * returns a vector of entities that have a damage hitbox collision between the vector of entities
+ * and the entity you pass in 
+ */
+std::vector<Entity *> CollisionHandler::detectMeleeCollision(std::vector<Entity*> returnObjects, const Entity *entity, const HitBox hb){
```
Comment (JohnAgapeyev) [inline src/collision/CollisionHandler.cpp]: Wrap it up

File: src/collision/CollisionHandler.cpp [hunk 7]
```diff
@@ -172,6 +172,26 @@ void CollisionHandler::detectLineCollision(TargetList &targetList, const int gun
...
+    std::vector<Entity*> allEntities;
+    for (const auto& obj: returnObjects){
+        if (obj != nullptr && entity != obj
+            && SDL_HasIntersection(&hb.getRect(), &obj->getDamHitBox().getRect())){
```
Comment (JohnAgapeyev) [inline src/collision/CollisionHandler.cpp]: Fix indentation

File: src/inventory/weapons/MeleeWeapon.cpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+    SDL_Rect meleeBox = {hitBoxX, hitBoxY, hitBoxW, hitBoxH};
+    HitBox hitBox(meleeBox);
+
+    CollisionHandler &ch = GameManager::instance()->getCollisionHandler();
+
+    //vector of each entity a zombie needs to check
+    std::vector<Entity *> hitMarines;
+    std::vector<Entity *> hitTurrets;
+    std::vector<Entity *> hitBarricades;
```
Comment (JohnAgapeyev) [inline src/inventory/weapons/MeleeWeapon.cpp]: Variables should be declared near where they are used
In this case, move these declarations down to include their assignments further down

**Review by JohnAgapeyev — CHANGES_REQUESTED** [review]

File: src/inventory/weapons/MeleeWeapon.cpp [hunk 9]
```diff
@@ -21,9 +21,10 @@
  * Description:
  * Ctor for MeleeWeapon
  */
-MeleeWeapon::MeleeWeapon(string type, TEXTURES sprite, string fireSound, string hitSound, string reloadSound,
-        string emptySound, int range, int damage, int AOE, int penetration, int clip,
-        int clipMax, int ammo, int reloadDelay, int fireDelay, int32_t id)
+MeleeWeapon::MeleeWeapon(const string& type, TEXTURES sprite, const string& fireSound,
+        const string& hitSound, const string& reloadSound, const string& emptySound, int range,
+        int damage, int AOE, int penetration, int clip, int clipMax, int ammo, int reloadDelay,
+        int fireDelay, int32_t id)
```
Comment (isaacmorneau) [inline src/inventory/weapons/MeleeWeapon.cpp]: The rest of the should be const as well

**Comment by JohnAgapeyev** [comment]
You've got merge conflicts yo

File: src/collision/CollisionHandler.cpp [hunk 10]
```diff
@@ -172,6 +172,27 @@ void CollisionHandler::detectLineCollision(TargetList &targetList, const int gun
...
+ *      std::vector<Entity*> returnObjects, const Entity *entity, const HitBox hb)
+ * Description:
+ * returns a vector of entities that have a damage hitbox collision between the vector of entities
+ * and the entity you pass in
+ */
+std::vector<Entity *> CollisionHandler::detectMeleeCollision(std::vector<Entity*> returnObjects,
+        const Entity *entity, const HitBox hb){
+    std::vector<Entity*> allEntities;
+    for (const auto& obj: returnObjects){
+        if (obj != nullptr && entity != obj
```
Comment (isaacmorneau) [inline src/collision/CollisionHandler.cpp]: `if (obj && ...` no need for nullptr check

File: src/inventory/weapons/MeleeWeapon.cpp [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+ * Author: Mark Tattrie
+ * Function Interface: MeleeWeapon::MeleeWeapon(string type, TEXTURES sprite, string fireSound,
+ *       string hitSound, string reloadSound, string emptySound, int range, int damage, int AOE,
+ *       int penetration, int clip, int clipMax, int ammo, int reloadDelay, int fireDelay, int32_t id)
+ * Description:
+ * Ctor for MeleeWeapon
+ */
+MeleeWeapon::MeleeWeapon(const string& type, TEXTURES sprite, const string& fireSound,
+        const string& hitSound, const string& reloadSound, const string& emptySound, int range,
+        int damage, int AOE, int penetration, int clip, int clipMax, int ammo, int reloadDelay,
```
Comment (isaacmorneau) [inline src/inventory/weapons/MeleeWeapon.cpp]: the rest of the params should also be const

File: src/inventory/weapons/ZombieHand.h [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    constexpr int AOE               = 1;
+    constexpr int PENETRATION       = 1;
+
+    constexpr int CLIP              = 10000;
+    constexpr int CLIPMAX           = 10000;
+    constexpr int AMMO              = 10000;
+
+    constexpr int RELOADDELAY       = 0;
+    constexpr int FIREDELAY         = 600;
+}
```
Comment (isaacmorneau) [inline src/inventory/weapons/ZombieHand.h]: these are variables that shouldnt change but please avoid space aligning like this as per the guide

File: src/inventory/weapons/ZombieHand.h [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+class ZombieHand: public MeleeWeapon {
+public:
+
+    ZombieHand(int32_t id);
```
Comment (isaacmorneau) [inline src/inventory/weapons/ZombieHand.h]: const

File: src/player/Marine.cpp [hunk 14]
```diff
@@ -21,6 +19,11 @@ void Marine::onCollision() {
 
 void Marine::collidingProjectile(const int damage) {
     health -= damage;
+    /**
+    if (health <= 0){
+        GameManager::instance()->deleteMarine(getId());
+    }
+    */
```
Comment (isaacmorneau) [inline src/player/Marine.cpp:27]: is this supposed to be commented?

File: src/turrets/Turret.cpp [hunk 15]
```diff
@@ -134,14 +134,19 @@ void Turret::activateTurret() {
...
  * Function Interface: void collidingProjectile(const int damage)
  * Description:
  * Damages the turret by the amount of 'damage' parameter passed in.
  */
 void Turret::collidingProjectile(const int damage) {
     health -= damage;
+    //turret health is less then or equal to 0
+    if(!healthCheckTurret()){
+        GameManager::instance()->deleteTurret(getId());
+    }
```
Comment (isaacmorneau) [inline src/turrets/Turret.cpp:149]: i ask about marine because this remove is in

**Review by isaacmorneau — CHANGES_REQUESTED** [review]

**Review by eyyu — APPROVED** [review]

File: src/inventory/weapons/MeleeWeapon.cpp [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+    const int hitBoxH = range;
+
+    SDL_Rect meleeBox = {hitBoxX, hitBoxY, hitBoxW, hitBoxH};
+    HitBox hitBox(meleeBox);
+
+    CollisionHandler &ch = GameManager::instance()->getCollisionHandler();
+
+    //vector of each marine the melee attack connects with
+    std::vector<Entity *> hitMarines;
+    hitMarines = ch.detectMeleeCollision(ch.getQuadTreeEntities(ch.quadtreeMarine, &mov),&mov, hitBox);
```
Comment (JohnAgapeyev) [inline src/inventory/weapons/MeleeWeapon.cpp]: Declaration and initialization can be done on the same line

**Review by JohnAgapeyev — APPROVED** [review]
A couple lines need condensing, but otherwise fine

**Review by isaacmorneau — APPROVED** [review]

**Comment by isaacmorneau** [comment]
once merge conflicts are fixed its good

**Pull request merged**